### PR TITLE
Add AssertionDependency

### PR DESCRIPTION
**Pull request by john-flanagan**
Following up on our discussion [here](https://github.com/tgrapperon/swift-dependencies-additions/discussions/17) about creating an `assert` dependency.

### CODE REVIEW

File: Sources/AssertionDependency/AssertionFailure.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import Dependencies
+import XCTestDynamicOverlay
+
+extension DependencyValues {
+  /// A ``AssertionFailureAction`` that will perform an anction when reached.
```
Comment (tgrapperon) [inline Sources/AssertionDependency/AssertionFailure.swift:5]:
```suggestion
  /// A ``AssertionFailureAction`` that will perform an action when reached.
```

File: Sources/AssertionDependency/AssertionFailure.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    // no-op
+  }
+
+  public static let testValue = AssertionFailureAction { message, file, line in
+    XCTFail(message(), file: file, line: line)
+  }
+}
+
+
+public struct AssertionFailureAction {
```
Comment (tgrapperon) [inline Sources/AssertionDependency/AssertionFailure.swift:28]:
```suggestion
public struct AssertionFailureAction: Sendable {
```

File: Sources/AssertionDependency/Assert.swift [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  public static let testValue = AssertAction { condition, message, file, line in
+    if !condition() {
+      XCTFail(message(), file: file, line: line)
+    }
+  }
+}
+
+public struct AssertAction {
```
Comment (tgrapperon) [inline Sources/AssertionDependency/Assert.swift:29]:
```suggestion
public struct AssertAction: Sendable {
```

File: Tests/AssertionDependencyTests/AssertionDependencyTests.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import AssertionDependency
+import Dependencies
+import XCTest
+
+final class AssertionDependencyTests: XCTestCase {
```
Comment (tgrapperon) [inline Tests/AssertionDependencyTests/AssertionDependencyTests.swift:5]: `XCTExpectFailure` doesn't work on Linux, so the (implicit) convention right now is to wrap these tests in `#if os(macOS)` (at least), so they run in CI.

File: Package.swift [hunk 5]
```diff
@@ -90,6 +91,19 @@ let package = Package(
       ]
     ),
 
+    .target(
+      name: "AssertionDependency",
```
Comment (tgrapperon) [inline Package.swift:95]: We need to `@_exported import AssertDependency` in the [`Exports.swift`](https://github.com/tgrapperon/swift-dependencies-additions/blob/main/Sources/DependenciesAdditions/Exports.swift) umbrella too.

**Review by tgrapperon — CHANGES_REQUESTED** [review]
Hey @john-flanagan! Thanks for the PR!
This is great! I was initially thinking about defining `assert` on top of `assertionFailure`, but it turns out that the [implementation](https://github.com/apple/swift/blob/2fc7659ed7874d069132add975df1178f36d0168/stdlib/public/core/Assert.swift#L40-L53) is very slightly different, so I guess it is better to keep them separated. We can maybe find a way to bundle both using some tricks involving keypaths, and overriding one would override both. I'll think about it, but this maybe not a good idea. In any case, this would be the subject of a dedicated PR.

**Comment by tgrapperon** [comment]
@john-flanagan, thanks for the PR! I'll cut a release in a few days.

**Comment by Sajjon** [comment]
@john-flanagan @tgrapperon cool, I just used this, very useful! I find myself in need of a `fatalError` Dependency as well, did any of you start working on that?

Maybe this lib can act as a source of inspiration https://github.com/SlaunchaMan/ErrorAssertions ? If you haven't started I might give it a go, all tips and recommendation would be welcome ofc :)

**Comment by tgrapperon** [comment]
@Sajjon Sorry, I missed your comment. 

I chose to only provide a dependency for `assert` and not for `precondition` or `fatalError` for security concerns. Users that are willing to crash should have their app crash immediately. In these case, the app is likely in an unexpected state. Providing a dependency for crashing would allow a third party to craft some values where the crash would be slightly deferred, and the breached app state logged somewhere for example.

I initially considered the same for assertions, but they are noop in `Release` mode, which is not the case of preconditions or fatal errors.

So that's why I'm a little wary of allowing such a dependency in the library for now. But feel free to open a discussion if you have use-cases or examples!

**Pull request merged**